### PR TITLE
Use `Iterable` instead `Iterator` on `iterate_in_threadpool`

### PR DESCRIPTION
**Pull request by jamesbraza**
# Summary

Having `AsyncContentStream` match typing of `SyncContentStream`, they should both be `Iterator` or `Iterable`, not separate.

~Given `iterate_in_threadpool` is typed with `Iterator`, and returns an `AsyncIterator`, I think then `AsyncContentStream` should be an `Iterator` as well~

From `mypy` inferred types in `test_streaming_response_custom_iterable`, it seems `AsyncIterable` is required. So, this PR standardizes on `Iterable`/`AsyncIterable`

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.

### CODE REVIEW

**Comment by jamesbraza** [comment]
Okay, nevermind, per type checks looks like `iterate_in_threadpool` and everything else should be updated to `AsyncIterable`. Will revisit this tomorrow

**Comment by jamesbraza** [comment]
Cc @tomchristie if you got time to take a look at this one, it's another bug I encountered recently

File: starlette/concurrency.py [hunk 1]
```diff
@@ -50,8 +50,8 @@ def _next(iterator: typing.Iterator[T]) -> T:
 
 
 async def iterate_in_threadpool(
-    iterator: typing.Iterator[T],
-) -> typing.AsyncIterator[T]:
+    iterator: typing.Iterable[T],
+) -> typing.AsyncIterable[T]:
     while True:
         try:
             yield await anyio.to_thread.run_sync(_next, iterator)
```
Comment (Kludex) [inline starlette/concurrency.py]: But `_next()` receives an Iterator... 👀

**Review by Kludex — CHANGES_REQUESTED** [review]

**Comment by lovelydinosaur** [comment]
@jamesbraza you probably want to check in with other encode maintainers for this project. (I'm focused on `httpx`, `httpcore` at the moment.)

File: starlette/concurrency.py (hunk 1 above)
Comment (jamesbraza) [reply starlette/concurrency.py]: I guess `mypy` didn't realize this, great catch! Looks like https://github.com/encode/starlette/pull/2357 updates to latest `mypy`, but unfortunately `scripts/check` still doesn't catch this

Comment (jamesbraza) [reply starlette/concurrency.py]: Alright the fix was simple, having `iterate_in_threadpool` call `iter` before using `_next`. PR is ready for review again

File: starlette/concurrency.py [hunk 2]
```diff
@@ -50,8 +50,9 @@ def _next(iterator: typing.Iterator[T]) -> T:
 
 
 async def iterate_in_threadpool(
-    iterator: typing.Iterator[T],
-) -> typing.AsyncIterator[T]:
+    iterable: typing.Iterable[T],
```
Comment (Kludex) [inline starlette/concurrency.py]: This is a breaking change.

File: starlette/concurrency.py [hunk 3]
```diff
@@ -50,8 +50,9 @@ def _next(iterator: typing.Iterator[T]) -> T:
...
+) -> typing.AsyncIterable[T]:
+    iterator = iter(iterable)
```
Comment (Kludex) [inline starlette/concurrency.py]: Can you create a test case that fails without this?

**Review by Kludex — CHANGES_REQUESTED** [review]

File: starlette/concurrency.py (hunk 2 above)
Comment (jamesbraza) [reply starlette/concurrency.py]: Okay gotchu. Would you like me to leave the name as `iterator`, and just update the type hint?

File: starlette/concurrency.py (hunk 3 above)
Comment (jamesbraza) [reply starlette/concurrency.py]: Sure, I can cook something up 👍

Comment (jamesbraza) [reply starlette/concurrency.py]: This test fails on `master` but passes here

```python
@pytest.mark.anyio
async def test_iterate_in_threadpool() -> None:
    class CustomIterable:
        def __iter__(self):
            yield from range(3)

    assert [v async for v in iterate_in_threadpool(CustomIterable())] == [0, 1, 2]
```

Let me know if you were hoping for something different

File: starlette/concurrency.py (hunk 2 above)
Comment (Kludex) [reply starlette/concurrency.py]: Yes, please.

File: starlette/concurrency.py [hunk 4]
```diff
@@ -50,8 +50,9 @@ def _next(iterator: typing.Iterator[T]) -> T:
 
 
 async def iterate_in_threadpool(
-    iterator: typing.Iterator[T],
-) -> typing.AsyncIterator[T]:
+    iterable: typing.Iterable[T],
+) -> typing.AsyncIterable[T]:
```
Comment (Kludex) [inline starlette/concurrency.py]: Isn't the return an `AsyncIterator`?

Comment (jamesbraza) [reply starlette/concurrency.py]: Good catch, thanks. Implemented this just now

File: starlette/concurrency.py (hunk 2 above)
Comment (jamesbraza) [reply starlette/concurrency.py]: Alright, renamed back to `iterator`

**Review by Kludex — APPROVED** [review]

**Pull request merged**